packages/affiliates/libs/uniswap/models.ts

```diff
@@ -0,0 +1,169 @@
+import assert from "assert";
+import { exists, getPositionKey } from "./utils";
+import { ethers } from "ethers";
+
+type Id = string | number;
+type HasId = {
+  id?: Id;
+};
+interface MakeId<T> {
+  (obj: T): string;
+}
+
+// Generic table with common functions, store is a map now, but we could replace this with any kind of
+// data store with some minor modifications, as long as it has, set, get, delete, has and values
+export function Table<T extends HasId>(config: { makeId: MakeId<T>; type: string }, store: Map<Id, T>) {
+  const { makeId, type } = config;
+  async function create(data: T) {
+    const id = makeId(data);
+    assert(!(await has(id)), `${type} exists`);
+    return set({ id, ...data });
+  }
+  async function set(data: T) {
+    assert(data.id, `${type} requires id`);
+    await store.set(data.id, { ...data });
+    return data;
+  }
+  async function get(id: string) {
+    assert(await store.has(id), `${type} does not exist`);
+    return store.get(id) as T;
+  }
+  async function has(id: string) {
+    return store.has(id);
+  }
+  async function list() {
+    return [...store.values()];
+  }
+  async function forEach(cb: (value: T, index: number, array: T[]) => void) {
+    const all = await list();
+    all.forEach(cb);
+  }
+  async function update(id: string, data: any = {}) {
+    const got = await get(id);
+    return set({ ...got, ...data });
+  }
+  return {
+    create,
+    set,
+    get,
+    has,
+    update,
+    forEach,
+    list,
+    store
+  };
+}
+
+export type NftPosition = {
+  id?: string;
+  tokenId: string;
+  amount?: string;
+  amount0?: string;
+  amount1?: string;
+  token0?: string;
+  token1?: string;
+  liquidity?: string;
+  feeGrowthInside0LastX128?: string;
+  feeGrowthInside1LastX128?: string;
+  tokensOwed0?: string;
+  tokensOwed1?: string;
+  pool?: string;
+};
+
+// Nft positions are gathered from the nft contract, and require slightly different indexing than the positions in pool contract
+export const NftPositions = () => {
```

Comment:
If you overload the makeId function, you might be able to avoid this function entirely (although it may not be cleaner).
```ts
// 
function makeId(data: NftPosition, type: "NftPositions");
function makeId(data: Position, type: "Positions");
function makeId(data: any, type: string) {
  if (type === "NftPositions") {
    // Typescript should (?) know that data is an NftPosition.
  else if (type === "Positions") {
    // Same for Positions.
  }
```

https://www.typescriptlang.org/docs/handbook/2/functions.html#function-overloads. Note: the implementation function is not considered a "valid" signature, meaning that you can't call it directly -- you have to use one of the overloads -- https://www.typescriptlang.org/docs/handbook/2/functions.html#overload-signatures-and-the-implementation-signature.

Not sure if you want to do it this way, but it might mean you can just define this overload and call it within the constructor of the Table and then not have to have any other code that's specific to these interfaces.

Reply:
interesting technique.... that is really cool. I think though in this case i want to keep the implementation of makeId encapsulated with the table, so they can be broken out as independent features.  will have to think more about how to use this though, i could see it being really useful in some situations.